pgjdbc/src/main/java/org/postgresql/core/ProtocolVersion.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2025, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.core;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.sql.SQLException;
+import java.util.Objects;
+
+public class ProtocolVersion {
+  private final int major;
+  private final int minor;
+
+  private ProtocolVersion() {
+    // not meant to be instantiated
+    major = 0;
+    minor = 0;
+  }
+
+  public ProtocolVersion(int major, int minor) {
+    this.major = major;
+    this.minor = minor;
+  }
+
+  public ProtocolVersion(int protocol) {
+    this.minor = protocol & 0xff;
+    this.major = (protocol >> 16 ) & 0xff;
+  }
```

Comment:
Please prefer static factory methods over constructors as they yield extra name.
I'm not sure `(int protocol)` constructor belongs here at it is an implementation detail of `NegotiateProtocolVersion` message

```suggestion
  private ProtocolVersion(int major, int minor) {
    this.major = major;
    this.minor = minor;
  }
  
  public static ProtocolVersion fromMajorMinor(int major, int minor) {
    if (major == 3 && minor == 0) {
      return V_3_0;
    }
    if (major == 3 && minor == 2) {
      return V_3_2;
    }
    return new ProtocolVersion(major, minor);
  }
```